src/wasm-interpreter.h

```diff
@@ -1469,17 +1469,54 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
     assert(cast.outcome == cast.Success);
     return cast.castRef;
   }
-  Flow visitBrOnCast(BrOnCast* curr) {
-    NOTE_ENTER("BrOnCast");
-    auto cast = doCast(curr);
-    if (cast.outcome == cast.Break) {
-      return cast.breaking;
+  Flow visitBrOn(BrOn* curr) {
+    NOTE_ENTER("BrOn");
+    // BrOnCast uses the casting infrastructure, so handle it first.
+    if (curr->op == BrOnCast) {
+      auto cast = doCast(curr);
+      if (cast.outcome == cast.Break) {
+        return cast.breaking;
+      }
+      if (cast.outcome == cast.Null || cast.outcome == cast.Failure) {
+        return cast.originalRef;
+      }
+      assert(cast.outcome == cast.Success);
+      return Flow(curr->name, cast.castRef);
+    }
+    // The others do a simpler check for the type.
+    Flow flow = visit(curr->ref);
+    if (flow.breaking()) {
+      return flow;
     }
-    if (cast.outcome == cast.Null || cast.outcome == cast.Failure) {
-      return cast.originalRef;
+    const auto& value = flow.getSingleValue();
+    NOTE_EVAL1(value);
+    // Prepare to break. If we do not break, we will unset the break target, and
+    // we just flow out the value.
+    flow = Flow(curr->name, value);
+    if (value.isNull()) {
+      flow.breakTo = Name();
```

Comment:
```suggestion
      flow.breakTo.clear();
```
Nit: I guess this reads more intuitive..? The same for similar lines